### PR TITLE
Graphite support

### PR DESCRIPTION
**Pull request by danielbeardsley**
Add support for writing the same metrics to graphite. Graphite now supports labels, so let's use those too.

Prometheus is better for real time alerting, while graphite is better for long term archival storage. So, let's support both (based on env vars)

### CODE REVIEW

File: lib/graphite.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import type { Labels, Metric } from "./metrics.js";
+
+export function getGraphiteBodyForMetrics(metrics: Metric[]): string {
```
Comment (djmetzle) [inline lib/graphite.ts:3]: Is there not a library we can use for building and sending messages to graphite? This module has multiple regexes and interpolations for building the statsd message format, but seemingly no tests? This seems like it ought to be a solved problem.

**Comment by kthaler** [comment]
Was able to send Alertra metrics to Graphite by doing the following:

First, I showed that Alertra is successfully displaying metrics:
- Cloned this repo
- Set the `ALERTRA_API_KEY` environment variable with the API key proved by @danielbeardsley 
- Run `/npm install` in the `alertra-prometheus-exporter` directory
- Then run `\npm run start`
- In a second terminal, navigate to this directory, run  `curl localhost:13964 | less`
  - This will populate the terminal with metrics fetched from Alertra 

Then I had Alertra send metrics to graphite:
- Checkout this branch in both terminals
- Ensure that the `ALERTRA_API_KEY` environment variable is set in both
- In one terminal run `nc -l 0.0.0.0 2003`
  - Graphite will begin listening for metrics
- In the other terminal, run `GRAPHITE_HOST=127.0.0.1 GRAPHITE_INTERVAL_S=5 \npm run start`
  - This populates the other terminal with the metrics sent from Alertra to Graphite.
  - Data is sent in the format `<metric path> <metric value> <metric timestamp>` where the metric path contains a set of tags separated by a `;`. 
  - Example:
Screenshot 2023-11-02 at 5 00 42 PM

QA 💎

File: lib/graphite.ts (hunk 1 above)
Comment (danielbeardsley) [reply lib/graphite.ts:3]: Yes, there are. Here's the most popular one: https://www.npmjs.com/package/graphite but it doesn't do anything to make names and labels safe and thus we'd still need to do the regexing.

Also, which API is more likely to change over time and thus require updates? An npm package, or the graphite api? I think in this case, the graphite API will be way more stable (looks like it's been backward compatible for more than a decade). 

If the protocol were complex, I'd vote for using someone else's code, but I think any other library would require more code and more upkeep and have more bugs.

It's a good point about a lack of tests and I don't have a great reason other than  I don't want to spend the time right now and I expect churn to be low enough (on the graphite protocol code) for tests to have less value.

Comment (djmetzle) [reply lib/graphite.ts:3]: Okay, thats fair, but i hope you understand the spirit of that critique here. We're implementing a wire protocol, without an established spec or tests. 
Is the standard implementation not stable?
image

Comment (danielbeardsley) [reply lib/graphite.ts:3]: > We're implementing a wire protocol, without an established spec or tests

Yes, there are no tests, but the protocol and its docs are the spec.

In general, you are right that we should prefer libraries that hide away some of the complexity of protocols so we don't spend our time creating and maintaining a less featureful, more buggy version of the same.

In this case, the most popular package doesn't remove the complexity (we'd still need to manually encode all the identifiers) and effectively would only replace ~10-15 lines of code at the expense of using an API and a package that has much greater chance of requiring changes or updates down the line.

**Comment by sterlinghirsh** [comment]
CR :zap:

**Pull request merged**